cookbook/controller/argument_value_resolver.rst

```diff
@@ -0,0 +1,208 @@
+    given argument. ``resolve()`` will only be executed when this returns ``true``.
+``resolve()``
+    This method will resolve the actual value for the argument. Once the value
+    is resolved, you should `yield`_ the value to the ``ArgumentResolver``.
+
+Both methods get the ``Request`` object, which is the current request, and an
+:class:`Symfony\\Component\\HttpKernel\\ControllerMetadata\\ArgumentMetadata`.
+This object contains all information retrieved from the method signature for
+the current argument.
+
+.. note::
+
+    The ``ArgumentMetadata`` is a simple data container created by the
+    :class:`Symfony\\Component\\HttpKernel\\ControllerMetadata\\ArgumentMetadataFactory`.
+    This factory will work on every supported PHP version but might give
+    different results. E.g. the ``isVariadic()`` will never return true on
+    PHP 5.5 and only on PHP 7.0 and higher it will give you basic types when
+    calling ``getType()``.
+
+Now that you know what to do, you can implement this interface. To get the
+current ``User``, you need the current security token. This token can be
+retrieved from the token storage::
+
+    namespace AppBundle\ArgumentValueResolver;
+
+    use AppBundle\Entity\User;
+    use Symfony\Component\HttpKernel\Controller\ArgumentValueResolverInterface;
+    use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+
+    class UserValueResolver implements ArgumentValueResolverInterface
+    {
+        private $tokenStorage;
+
+        public function __construct(TokenStorageInterface $tokenStorage)
+        {
+            $this->tokenStorage = $tokenStorage;
+        }
+
+        public function supports(Request $request, ArgumentMetadata $argument)
+        {
+            if (User::class !== $argument->getType()) {
+                return false;
+            }
+
+            $token = $this->tokenStorage->getToken();
+
+            if (!$token instanceof TokenInterface) {
+                return false;
+            }
+
+            return $token->getUser() instanceof User;
+        }
+
+        public function resolve(Request $request, ArgumentMetadata $argument)
+        {
+            yield $this->tokenStorage->getToken()->getUser();
+        }
+    }
+
+In order to get the actual ``User`` object in your argument, the given value
+should fulfill the following requirements:
+
+* The argument type (of the method signature) must be typehinted as ``User``;
+* The security token must be present;
+* The value should be an instance of the ``User``.
+
+When all those requirements are met and true is returned, the ``ArgumentResolver``
+calls ``resolve()`` with the same values as it called ``supports()``.
+
+.. tip::
+
+    You can leverage the ``DefaultValueResolver`` by making your resolver
+    accept only mandatory arguments. Given your signature is `User $user = null`,
+    the above example will not hit ``resolve()`` as one of the conditions
```

Comment:
I don't understand what condition is not matched. Do you mean that the resolving is done by `DefaultValueResolver` because `null` is not an instance of `User` or because the argument is not mandatory? The second case is not explicit in the conditions list.

Your example in the docs above **uses** a mandatory signature of the user, so in fact I don't understand where is the leverage. In that case the default value resolver is not called, but if the token storage returns no user, the user resolver is not called either (what should end with a php error for a missing parameter?).

Actually you user resolver cannot return `null` since you test it in `supports()`, so I wonder what's happening if the argument is defaulted, you say here that the default value resolver is called instead of the user resolver, but maybe the last test of `supports()` is not needed to let the token storage return `null` but then it should be tested in `resolves()` by `$argument->isOptional()` and throw a proper exception when the user cannot be retrieved whereas it is mandatory?

I feel like I miss something...